CDCgov/prime-simplereport
# upgrade okta to 12.0.2

# BACKEND PULL REQUEST

## Related Issue

- Resolves #5065 
- Resolves #6190

## Changes Proposed

- Upgrade Okta SDK to 12
  - With the new SDK, different api clients are now needed for each type of object. For example, the UserAPI is used to access/modify a user and the UserFactorAPI is used to access/modify a user's factors(mfa).
  - `ResourceException` was replaced with `APIException`. If the exception is from a call to okta, the [responseBody](https://github.com/CDCgov/prime-simplereport/pull/6125/files#diff-3b8491ab272d4e67c9efe508d5cc32d357709869e5af4c937bd03683d0228e6aR72) will be a json string version of the HTTP response. The `ApiExceptionHelper.getError` is the helper method to parse that response.
 

## Additional Information

- E2E Test: Security questions were updated in order to pass the wiremock matching.
- `resendActivationPasscode` and `enrollSecurityKey` was refactored to use the newly methods exposed in the sdk rather than making http requests.
- Existing "bugs" with okta (not a regression based on this update): 
  - Adding a user to too many facilities fail #6214
  - Enrolling security key as MFA fails #6215 
 

## Testing

- Wiremocks mapping were deleted and recreated which creates a lot of files changed. Use the [filters to deselect the .json files](https://github.com/CDCgov/prime-simplereport/pull/5365/files?file-filters%5B%5D=.gradle&file-filters%5B%5D=.java&file-filters%5B%5D=.lockfile&show-deleted-files=true&show-viewed-files=true). 
- Create an organization
  - Manually verify organization
  - Set up user via the UAC flow from the email link. Make sure you change the URL from the email from `test` to `dev2`
  - Add a facility
  - Add a user 
  - Delete a user
  - Reset user password
  - Resend user activation email
  - Reset MFA
  - Change user facilities
